sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1585,3 +1586,110 @@ object ResolveUpCast extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Maps a time column to multiple time windows using the Expand operator. Since it's non-trivial to
+ * figure out how many windows a time column can map to, we over-estimate the number of windows and
+ * filter out the rows where the time column is not inside the time window.
+ */
+object TimeWindowing extends Rule[LogicalPlan] {
+
+  /**
+   * Depending on the operation, the TimeWindow expression may be wrapped in an Alias (in case of
+   * projections) or be simply by itself (in case of groupBy),
+   * @param f The function that we want to apply on the TimeWindow expression
+   * @return The user defined function applied on the TimeWindow expression
+   */
+  private def getWindowExpr[E](f: TimeWindow => E): PartialFunction[Expression, E] = {
```

Comment:
Usually we would do this as an extractor (an unapply method). I think it looks a little less magical that way.

However, I'm worried that we are loosing the alias.  Can you add a test like:

```scala
df.groupBy(window(...).as("time")).agg(count("*")).select($"time")
```